File: PyTorch/utilities/deepWB.py
"""
 Deep white-balance editing main function (inference phase)
 Copyright (c) 2019 Samsung Electronics Co., Ltd. All Rights Reserved
 If you use this code, please cite the following paper:
 Mahmoud Afifi and Michael S Brown. Deep White-Balance Editing. In CVPR, 2020.
"""
__author__ = "Mahmoud Afifi"
__credits__ = ["Mahmoud Afifi"]

import numpy as np
import torch
from torchvision import transforms
import utilities.utils as utls


def deep_wb(image, task='all', net_awb=None, net_t=None, net_s=None, device='cpu', s=656):
    # check image size
    image_resized = image.resize((round(image.width / max(image.size) * s), round(image.height / max(image.size) * s)))
    w, h = image_resized.size
    if w % 2 ** 4 == 0:
        new_size_w = w
    else:
        new_size_w = w + 2 ** 4 - w % 2 ** 4

    if h % 2 ** 4 == 0:
        new_size_h = h
    else:
        new_size_h = h + 2 ** 4 - h % 2 ** 4

    inSz = (new_size_w, new_size_h)
    if not ((w, h) == inSz):
        image_resized = image_resized.resize(inSz)

    image = np.array(image)
    image_resized = np.array(image_resized)
    img = image_resized.transpose((2, 0, 1))
    img = img / 255
    img = torch.from_numpy(img)
    img = img.unsqueeze(0)
    img = img.to(device=device, dtype=torch.float32)

    if task == 'all':
        net_awb.eval()
        net_t.eval()
        net_s.eval()
        with torch.no_grad():
            output_awb = net_awb(img)
            output_t = net_t(img)
            output_s = net_s(img)

        tf = transforms.Compose([
            transforms.ToPILImage(),
            transforms.ToTensor()
        ])

        output_awb = tf(torch.squeeze(output_awb.cpu()))
        output_awb = output_awb.squeeze().cpu().numpy()
        output_awb = output_awb.transpose((1, 2, 0))
        m_awb = utls.get_mapping_func(image_resized, output_awb)
        output_awb = utls.outOfGamutClipping(utls.apply_mapping_func(image, m_awb))

        output_t = tf(torch.squeeze(output_t.cpu()))
        output_t = output_t.squeeze().cpu().numpy()
        output_t = output_t.transpose((1, 2, 0))
        m_t = utls.get_mapping_func(image_resized, output_t)
        output_t = utls.outOfGamutClipping(utls.apply_mapping_func(image, m_t))

        output_s = tf(torch.squeeze(output_s.cpu()))
        output_s = output_s.squeeze().cpu().numpy()
        output_s = output_s.transpose((1, 2, 0))
        m_s = utls.get_mapping_func(image_resized, output_s)
        output_s = utls.outOfGamutClipping(utls.apply_mapping_func(image, m_s))

        return output_awb, output_t, output_s

    elif task == 'awb':
        net_awb.eval()
        with torch.no_grad():
            output_awb = net_awb(img)

        tf = transforms.Compose([
            transforms.ToPILImage(),
            transforms.ToTensor()
        ])

        output_awb = tf(torch.squeeze(output_awb.cpu()))
        output_awb = output_awb.squeeze().cpu().numpy()
        output_awb = output_awb.transpose((1, 2, 0))
        m_awb = utls.get_mapping_func(image_resized, output_awb)
        output_awb = utls.outOfGamutClipping(utls.apply_mapping_func(image, m_awb))

        return output_awb

    elif task == 'editing':
        net_t.eval()
        net_s.eval()
        with torch.no_grad():
            output_t = net_t(img)
            output_s = net_s(img)

        tf = transforms.Compose([
            transforms.ToPILImage(),
            transforms.ToTensor()
        ])

        output_t = tf(torch.squeeze(output_t.cpu()))
        output_t = output_t.squeeze().cpu().numpy()
        output_t = output_t.transpose((1, 2, 0))
        m_t = utls.get_mapping_func(image_resized, output_t)
        output_t = utls.outOfGamutClipping(utls.apply_mapping_func(image, m_t))

        output_s = tf(torch.squeeze(output_s.cpu()))
        output_s = output_s.squeeze().cpu().numpy()
        output_s = output_s.transpose((1, 2, 0))
        m_s = utls.get_mapping_func(image_resized, output_s)
        output_s = utls.outOfGamutClipping(utls.apply_mapping_func(image, m_s))

        return output_t, output_s

File: Matlab/utilities/deepWB.m
%% Deep white-balance editing main function (inference phase)
% Author: Mahmoud Afifi
% Copyright (c) 2019 Samsung Electronics Co., Ltd. All Rights Reserved
% Please cite our paper:
% Mahmoud Afifi and Michael S Brown. Deep White-Balance Editing. In CVPR, 2020.
%%


function [varargout] = deepWB(I,net,target,device, depth, S, reportTime, normalize)
warning off
if nargin == 2
    target = 'all';
    device = 'gpu';
    depth = 4;
    S = 656;
    reportTime = 0;
    normalize = 0;
elseif nargin == 3
    device = 'gpu';
    depth = 4;
    S = 656;
    reportTime = 0;
    normalize = 0;
elseif nargin == 4
    depth = 4;
    S = 656;
    reportTime = 0;
    normalize = 0;
elseif nargin == 5
    S = 656;
    reportTime = 0;
    normalize = 0;
elseif nargin == 6
    reportTime = 0;
    normalize = 0;
elseif nargin == 7
    normalize = 0;
end

I_resized = imresize(I,[round(size(I,1)/max(size(I))*S)...
    round(size(I,2)/max(size(I))*S)]);

sz = size(I_resized);

if mod(sz(1),2^depth) == 0
    new_size_1 = sz(1);
else
    new_size_1 = sz(1) + 2^depth - mod(sz(1),2^depth);
end

if mod(sz(2),2^depth) == 0
    new_size_2 = sz(2);
else
    new_size_2 = sz(2) + 2^depth - mod(sz(2),2^depth);
end

inSz = [new_size_1, new_size_2, 3];

if sum(sz == inSz) ~= 3
    I_resized = imresize(I_resized,[inSz(1) inSz(2)]);
end

if isfield(net,'AWB_net') == 0
    [AWB_net,T_net,S_net]  = splitNetworks(net, depth);
else
    AWB_net = net.AWB_net;
    T_net = net.T_net;
    S_net = net.S_net;
end

switch lower(target)
    case 'all'
        if reportTime == 1
            tic;
        end
        
        if strcmpi(device,'gpu') == 1  && canUseGPU
            I_resized = gpuArray(dlarray(I_resized,'SSC'));
            I_AWB = gather(extractdata(predict(AWB_net,I_resized)));
            I_T = gather(extractdata(predict(T_net,I_resized)));
            I_S = gather(extractdata(predict(S_net,I_resized)));
        else
            I_resized = dlarray(I_resized,'SSC');
            I_AWB = extractdata(predict(AWB_net,I_resized));
            I_T = extractdata(predict(T_net,I_resized));
            I_S = extractdata(predict(S_net,I_resized));
        end
        
        
        I_resized = imresize(I,[300,300]);
        I_AWB_resized = imresize(I_AWB,[300,300]);
        I_T_resized = imresize(I_T,[300,300]);
        I_S_resized = imresize(I_S,[300,300]);
        
        m_AWB = phi(reshape(I_resized,[],3))\...
            reshape(I_AWB_resized,[],3);
        I_AWB = reshape(phi(reshape(I,[],3)) * m_AWB,size(I));
        I_AWB = out_of_gamut_clipping(I_AWB);
        
        m_T = phi(reshape(I_resized,[],3))\...
            reshape(I_T_resized,[],3);
        I_T = reshape(phi(reshape(I,[],3)) * m_T,size(I));
        I_T = out_of_gamut_clipping(I_T);
        
        m_S = phi(reshape(I_resized,[],3))\...
            reshape(I_S_resized,[],3);
        I_S = reshape(phi(reshape(I,[],3)) * m_S,size(I));
        I_S = out_of_gamut_clipping(I_S);
        
        if normalize == 0
            varargout{1} = I_AWB; varargout{2} = I_T; varargout{3} = I_S;
        else
            I_AWB = I_AWB./(sqrt(sum(I_AWB.^2,3))) .* (sqrt(sum(I.^2,3)));
            varargout{1} = I_AWB;
            I_T = I_T./(sqrt(sum(I_T.^2,3))) .* (sqrt(sum(I.^2,3)));
            varargout{2} = I_T;
            I_S = I_S./(sqrt(sum(I_S.^2,3))) .* (sqrt(sum(I.^2,3)));
            varargout{3} = I_S;
        end
        if reportTime == 1
            toc
        end
        
    case 'awb'
        
        if reportTime == 1
            tic
        end
        if strcmpi(device,'gpu') == 1  && canUseGPU
            I_resized = gpuArray(dlarray(I_resized,'SSC'));
            I_AWB = gather(extractdata(predict(AWB_net,I_resized)));
        else
            I_resized = dlarray(I_resized,'SSC');
            I_AWB = extractdata(predict(AWB_net,I_resized));
        end
        
        if sum(sz==inSz) ~= 3
            I_AWB = imresize(I_AWB,[sz(1) sz(2)]);
        end
        
        I_resized = imresize(I,[300,300]);
        I_AWB_resized = imresize(I_AWB,[300,300]);
        
        m_AWB = phi(reshape(I_resized,[],3))\...
            reshape(I_AWB_resized,[],3);
        I_AWB = reshape(phi(reshape(I,[],3)) * m_AWB,size(I));
        I_AWB = out_of_gamut_clipping(I_AWB);
        
        
        if normalize == 0
            varargout{1} = I_AWB;
        else
            I_AWB = I_AWB./(sqrt(sum(I_AWB.^2,3))) .* (sqrt(sum(I.^2,3)));
            varargout{1} = I_AWB;
        end
        if reportTime == 1
            toc
        end
    case 'editing'
        if reportTime == 1
            tic
        end
        if strcmpi(device,'gpu') == 1  && canUseGPU
            I_resized = gpuArray(dlarray(I_resized,'SSC'));
            I_T = gather(extractdata(predict(T_net,I_resized)));
            I_S = gather(extractdata(predict(S_net,I_resized)));
        else
            I_resized = dlarray(I_resized,'SSC');
            I_T = extractdata(predict(T_net,I_resized));
            I_S = extractdata(predict(S_net,I_resized));
        end
        
        I_resized = imresize(I,[300,300]);
        I_T_resized = imresize(I_T,[300,300]);
        I_S_resized = imresize(I_S,[300,300]);
        
        
        m_T = phi(reshape(I_resized,[],3))\...
            reshape(I_T_resized,[],3);
        I_T = reshape(phi(reshape(I,[],3)) * m_T,size(I));
        I_T = out_of_gamut_clipping(I_T);
        
        m_S = phi(reshape(I_resized,[],3))\...
            reshape(I_S_resized,[],3);
        I_S = reshape(phi(reshape(I,[],3)) * m_S,size(I));
        I_S = out_of_gamut_clipping(I_S);
        
        if normalize == 0
            varargout{1} = I_T; varargout{2} = I_S;
        else
            
            I_T = I_T./(sqrt(sum(I_T.^2,3))) .* (sqrt(sum(I.^2,3)));
            varargout{1} = I_T;
            I_S = I_S./(sqrt(sum(I_S.^2,3))) .* (sqrt(sum(I.^2,3)));
            varargout{2} = I_S;
        end
        if reportTime == 1
            toc
        end
end

